Employ tracing extended basic blocks

qwe661234 opened the pull request:
We enhance the original EBB by introducing trace EBB. Trace EBB employs two main strategies: recursive jump translation and retroactive block chaining. After decoding instructions, we can determine the target addresses of jal, cj, and cjal instructions. This enables us to recursively translate the instructions of the jump target into the same basic block. As for the second strategy, we chain the previous block to the current block after emulation since the branch target can be obtained during emulation.

Based on performance results below, we observed improvements resulting from the trace EBB strategy.

|Test| commit 1c11b39 |trace EBB |Speedup|
|--------| -------- | -------- | -------- | 
| CoreMark | 1155.174 (Iterations/Sec)| 1351.065 (Iterations/Sec) |+16.64％|
| dhrystone  | 1017 DMIPS |1073 DMIPS |+5.5%|
| nqueens |8630 msec   | 8295 msec   |+4.0%|

jserv commented on src/emulate.c:
```diff
@@ -287,18 +287,21 @@ enum {
-            return true;                                                  \
-        const rv_insn_t *next = ir + 1;                                   \
-        MUST_TAIL return next->impl(rv, next);                            \
+/* tracing information for retroactive block chaining */
```
Improve the comments.

jserv commented on src/emulate.c:
```diff
@@ -356,107 +359,43 @@ RVOP(jalr, {
-    RV_EXC_MISALIGN_HANDLER(pc, insn, false, 0);
-    if (ir->branch_taken)
-        return ir->branch_taken->impl(rv, ir->branch_taken);
+#define BRNACH_FUNC(type, comp)                                  \
```
Typo. Should be `BRANCH_FUNC`.

jserv replied:
Rename `comp` to `cond`, which means "condition."

jserv commented on src/emulate.c:
```diff
@@ -287,18 +287,21 @@ enum {
-        const rv_insn_t *next = ir + 1;                                   \
-        MUST_TAIL return next->impl(rv, next);                            \
+/* tracing information for retroactive block chaining */
+static bool branch_take = false;
```
Should it be `branch_taken`?

jserv commented on src/emulate.c:
```diff
@@ -356,107 +359,43 @@ RVOP(jalr, {
+    branch_take = true;                                          \
+    rv->PC += ir->imm;                                           \
+    /* check instruction misaligned */                           \
+    RV_EXC_MISALIGN_HANDLER(pc, insn, false, 0);                 \
```
Is it possible to omit misalign handler for `BRANCH_FUNC`?

qwe661234 replied:
No, we just define macro `BRANCH_FUNC` to avoid duplicate code, the implementation deos not change.

jserv commented on src/emulate.c:
```diff
@@ -1447,27 +1397,41 @@ void rv_step(riscv_t *rv, int32_t cycles)
         assert(block);
 
+        /* retroactive block chaining */
+        if (prev) {
```
Illustrate the algorithm in comments.

jserv reviewed (changes requested):
Proofread the git commit messages as following:
> Enhancing the original extended basic block (EBB) with tracing EBB,
> employing two main strategies: recursive jump translation and
> retroactive block chaining.
> 
> Target addresses of jal, cj, and cjal instructions are determined
> after instruction decoding, allowing for recursive translation of
> jump target instructions into the same basic block. Additionally,
> the previous block is chained to the current block after emulation
> using the obtained branch target during emulation.
> 
> Performance improvements observed based on the results below,
> resulting from the tracing EBB strategy.

jserv commented on src/emulate.c:
```diff
@@ -1294,6 +1238,26 @@ static bool insn_is_branch(uint8_t opcode)
     return false;
 }
 
+static bool insn_is_unconditional_jump(uint8_t opcode)
```
The naming is a bit confusing. In RISC-V, these instructions are classified as "unconditional branch."

You shall refine the naming scheme of `insn_is_branch`.

jserv commented on src/emulate.c:
```diff
@@ -356,107 +360,43 @@ RVOP(jalr, {
-        return ir->branch_taken->impl(rv, ir->branch_taken);
+#define BRANCH_FUNC(type, cond)                                  \
+    const uint32_t pc = rv->PC;                                  \
+    if ((type) rv->X[ir->rs1] cond(type) rv->X[ir->rs2]) {       \
```
Should be
```c
if ((type) rv->X[ir->rs1] cond (type) rv->X[ir->rs2]) {
```

Add proper `/* clang-format off */` and `/* clang-format on */` for better indention.

qwe661234 committed:
Add Trace EBB

We enhance the original EBB by introducing trace EBB. Trace EBB employs
two main strategies: recursive jump translation and retroactive block chaining.
After decoding instructions, we can determine the target addresses of jal,
cj, and cjal instructions. This enables us to recursively translate the
instructions of the jump target into the same basic block. As for the second
strategy, we chain the previous block to the current block after emulation
since the branch target can be obtained during emulation.

Based on performance results below, we observed improvements resulting
from the trace EBB strategy.

|   Test   |      commit 1c11b39    |         trace EBB       |Speedup|
|----------+------------------------+-------------------------+-------|
| CoreMark |1155.174(Iterations/Sec)|1351.065 (Iterations/Sec)|+16.6% |
|----------+------------------------+-------------------------+-------|
| dhrystone|        1017 DMIPS      |        1073 DMIPS       | +5.5% |
|----------+------------------------+-------------------------+-------|
| nqueens  |        8630 msec       |        8295 msec        | +4.0% |
